### PR TITLE
Training utilities for custom training loops (progress bar and timer)

### PR DESCRIPTION
**Pull request by bkvogel**
Add some training utilities for #2252. This PR supports the progress bar functionality of `ProgressBar` to be used without requiring the usage of Trainer. A `TimerUtility` has also been added which allows the printing of training updates or other actions to be triggered at a specified rate, such as every x seconds for example. 

Since the existing `ProgressBar` requires a Trainer object but shared code with this progress bar, the common code has been moved into a new class, `ProgressBarPrinter`, to avoid code duplication.

Since I suspect the functionality and/or API is still likely to change, I have marked as experimental.

Other utilities such as those provided as trainer extensions can be added later as separate PRs.

The existing "custom training loop" examples for MNIST, CIFAR, and ptb have been modified to use either the new `TimerUtility` or the timer-based progress bar, `IteratorProgressBar`.

### CODE REVIEW

File: chainer/training/non_trainer_extensions/progress_utility.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+from __future__ import division
+import time
+
+from chainer.training.extensions import ProgressBarPrinter
+from chainer import utils
+
+
+class TimerUtility(object):
+
+    """Timer utility
```
Comment (unnonouno) [inline chainer/training/non_trainer_extensions/progress_utility.py]: Insert a period.

File: chainer/training/non_trainer_extensions/progress_utility.py [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+
+    This is a callable utility that will always return True if called at a
+    rate slower than the specified interval. However, if it is called at a
+    faster rate, it will only return True at a maximum rate of once every
+    `interval_seconds` seconds.
```
Comment (unnonouno) [inline chainer/training/non_trainer_extensions/progress_utility.py]: Use two backquotes for rst files.

File: chainer/training/non_trainer_extensions/progress_utility.py [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+        utils.experimental('chainer.training.non_trainer_extensions.TimerUtility')  # NOQA
+        self.iteration = 0
+        self.speed = 0
+
+        self._start_time = time.time()
+        self._interval_seconds = interval_seconds
+        self._last_time = time.time()
+
+    def __call__(self):
+        """Return True or False depending on elapsed time since last call.
```
Comment (unnonouno) [inline chainer/training/non_trainer_extensions/progress_utility.py]: Returns

File: chainer/training/non_trainer_extensions/progress_utility.py [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+
+        This method should be called once each iteration so that the
+        `iteration` and `speed` attributes will be updated appropriately.
```
Comment (unnonouno) [inline chainer/training/non_trainer_extensions/progress_utility.py]: two backquotes

File: chainer/training/non_trainer_extensions/progress_utility.py [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+
+        If at least `interval_seconds` has elapsed since the last call, return
```
Comment (unnonouno) [inline chainer/training/non_trainer_extensions/progress_utility.py]: ditto

File: chainer/training/non_trainer_extensions/progress_utility.py [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+        True. Otherwise return False.
+        """
+        self.iteration += 1
+        ret_val = False
+        last_time = self._last_time
+        now = time.time()
+        delta_time = now - last_time
+        self.speed = float(self.iteration) / (now - self._start_time)
+        if delta_time >= self._interval_seconds:
+            self._last_time = now
```
Comment (unnonouno) [inline chainer/training/non_trainer_extensions/progress_utility.py]: I want to discuss this implementation. I think `self._last_time += self._interval_seconds` triggers more precise timing. How do you think?

File: chainer/training/non_trainer_extensions/progress_utility.py [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+    Specifically, this utility keeps track of the progress within the
+    current epoch, when a new epoch is reached, and when
+    the end of the training is reached.
+
+    Usage:
+        This utility is intended to be used to display a progress bar
+        and track epoch/iteration progress in a cutom training loop
+        (that is not using Trainer) in which a dataset iterator is
+        not used. If a dataset iterator is used then
+        `IteratorProgressBar` may be more appropriate.
```
Comment (unnonouno) [inline chainer/training/non_trainer_extensions/progress_utility.py]: use <pre>:class:`IteratorProgressBar`</pre>

File: chainer/training/non_trainer_extensions/progress_utility.py [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+
+    Args:
+        iters_per_epoch (int): The number of iterations in one epoch.
+
```
Comment (unnonouno) [inline chainer/training/non_trainer_extensions/progress_utility.py]: Delete a blank line.

File: chainer/training/non_trainer_extensions/progress_utility.py [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+        training_length (tuple): Length of whole training. It consists of an
+            integer and either ``'epoch'`` or ``'iteration'``.
+
```
Comment (unnonouno) [inline chainer/training/non_trainer_extensions/progress_utility.py]: ditto

File: chainer/training/non_trainer_extensions/progress_utility.py [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+        enable_progress_bar (bool): If True, enable a progress bar display
+        that will be refreshed at a maximum rate of `interval_seconds`.
```
Comment (unnonouno) [inline chainer/training/non_trainer_extensions/progress_utility.py]: Indent?

File: chainer/training/non_trainer_extensions/progress_utility.py [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+
+        interval_seconds (float): This sets the maximum rate, in seconds, at
+        which this object will return True when called.
```
Comment (unnonouno) [inline chainer/training/non_trainer_extensions/progress_utility.py]: ditto

File: chainer/training/non_trainer_extensions/progress_utility.py [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+        self._training_length = training_length
+        self._enable_progress_bar = enable_progress_bar
+        self._progress_bar = ProgressBarPrinter(training_length)
+
+    def __call__(self):
+        """Return True or False depending on elapsed time since last call.
+
+        This method should be called once each iteration.
+
+        If at least `interval_seconds` has elapsed since the last call, return
```
Comment (unnonouno) [inline chainer/training/non_trainer_extensions/progress_utility.py]: ditto

File: chainer/training/non_trainer_extensions/progress_utility.py [hunk 13]
```diff
@@ -0,0 +1,199 @@
...
+        self.is_new_epoch = False
+
+        # private attribute:
+        self._iters_per_epoch = iters_per_epoch
+        self._training_length = training_length
+        self._enable_progress_bar = enable_progress_bar
+        self._progress_bar = ProgressBarPrinter(training_length)
+
+    def __call__(self):
+        """Return True or False depending on elapsed time since last call.
```
Comment (unnonouno) [inline chainer/training/non_trainer_extensions/progress_utility.py]: Returns

File: chainer/training/non_trainer_extensions/progress_utility.py [hunk 14]
```diff
@@ -0,0 +1,199 @@
...
+
+        If at least `interval_seconds` has elapsed since the last call, return
+        True. Otherwise return False.
+        """
+        if self.iteration % self._iters_per_epoch == 0 and self.iteration != 0:
+            self.is_new_epoch = True
+            self.epoch += 1
+        else:
+            self.is_new_epoch = False
+        self.epoch_detail = float(self.iteration) /\
```
Comment (unnonouno) [inline chainer/training/non_trainer_extensions/progress_utility.py]: Insert a space character before a backslash.

File: chainer/training/non_trainer_extensions/progress_utility.py [hunk 15]
```diff
@@ -0,0 +1,199 @@
...
+        return ret_val
+
+
+class IteratorProgressBar(TimerUtility):
+
+    """TimerUtility that displays a progress bar for a training loop.
+
+    This is a TimerUtility that displays a progress bar using information
+    from a supplied iterator. The progress bar display is updated at the
+    same rate that `__call__()` returns True.
```
Comment (unnonouno) [inline chainer/training/non_trainer_extensions/progress_utility.py]: ditto

File: chainer/training/non_trainer_extensions/progress_utility.py [hunk 16]
```diff
@@ -0,0 +1,199 @@
...
+
+    def __init__(self, iterator, training_length,
+                 interval_seconds=2):
+        super(IteratorProgressBar, self).__init__(interval_seconds)
+        self._iterator = iterator
+        self._training_length = training_length
+        self._progress_bar = ProgressBarPrinter(training_length)
+
+    def __call__(self):
+        """Return True or False depending on elapsed time since last call.
```
Comment (unnonouno) [inline chainer/training/non_trainer_extensions/progress_utility.py]: Returns.

File: chainer/training/non_trainer_extensions/progress_utility.py [hunk 17]
```diff
@@ -0,0 +1,199 @@
...
+
+        This method should be called once each iteration.
+        For the iterator that was supplied to the initializer, it is also
+        assumed that iterator.next() is called once per iteration and is
```
Comment (unnonouno) [inline chainer/training/non_trainer_extensions/progress_utility.py]: Use backquotes.

File: chainer/training/non_trainer_extensions/progress_utility.py [hunk 18]
```diff
@@ -0,0 +1,199 @@
...
+        always called before calling this method.
+
+        If at least `interval_seconds` has elapsed since the last call, return
```
Comment (unnonouno) [inline chainer/training/non_trainer_extensions/progress_utility.py]: ditto

File: examples/cifar/train_cifar_custom_loop.py [hunk 19]
```diff
@@ -100,20 +109,20 @@ def main():
...
-            model.predictor.train = False
-            for batch in test_iter:
-                x_array, t_array = convert.concat_examples(batch, args.gpu)
-                x = chainer.Variable(x_array)
-                t = chainer.Variable(t_array)
-                loss = model(x, t)
-                sum_loss += float(loss.data) * len(t.data)
-                sum_accuracy += float(model.accuracy.data) * len(t.data)
+            # It is good practice to turn off train mode during evaluation.
+            with configuration.using_config('train', False):
```
Comment (unnonouno) [inline examples/cifar/train_cifar_custom_loop.py]: Oh, is it a severe bug?

File: chainer/training/non_trainer_extensions/progress_utility.py [hunk 20]
```diff
@@ -0,0 +1,199 @@
...
+        delta_time = now - last_time
+        self.speed = float(self.iteration) / (now - self._start_time)
+        if delta_time >= self._interval_seconds:
+            self._last_time = now
+            ret_val = True
+
+        return ret_val
+
+
+class ProgressUtility(TimerUtility):
```
Comment (unnonouno) [inline chainer/training/non_trainer_extensions/progress_utility.py]: Do we really need to use inheritance here? Using a member variable is enough simple i feel. How do you think?

**Comment by bkvogel** [comment]
Thank you for the feedback and sorry for the delay. I am preparing a PR for ProgressBarPrinter and plan to submit soon. I am fine with `chainer.utils` if you think it is a suitable place to put it. I am not sure I understand your inheritance concern with `TimerUtility`, but I will think about it.

File: examples/cifar/train_cifar_custom_loop.py (hunk 19 above)
Comment (bkvogel) [reply examples/cifar/train_cifar_custom_loop.py]: Yes, without this, layers such as dropout will not work correctly in test mode. Since it may take a while longer for this PR, shall I submit a separate bug-fix PR?
